### PR TITLE
"How to add threads to your team chat" tutorial

### PR DESCRIPTION
**Pull request by keerlu**
Tutorial to go with the example code in https://github.com/talkjs/talkjs-examples/pull/548

Reviewing:

@csmeyns, as well as general writing stuff please could you keep an eye out for continuity errors/the flow not making sense? I've written it in lots of small chunks over quite a while so some weird ordering might have crept in.

@Julianerik, do you mind having a quick look (particularly at the back button section, which differs more from the old threads tutorial than the rest) and checking the code description parts are alright?

### CODE REVIEW

File: how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md [hunk 1]
```diff
@@ -0,0 +1,712 @@
+# How to add threads to your team chat with TalkJS
+
+This tutorial builds on [How to use TalkJS to create a team chat with channels](https://talkjs.com/resources/how-to-use-talkjs-to-create-a-team-chat-with-channels/), which shows you how to use TalkJS's [React SDK](https://talkjs.com/docs/Getting_Started/React/) to make a collaborative chat with DMs and discussion channels, similar to Slack or Teams.
+
+In this tutorial we'll show you how to add reply threads:
+
+!! 1-demo.gif
```
Comment (keerlu) [inline how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md:7]: Reminder to myself to add the image on Ghost

File: how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md [hunk 2]
```diff
@@ -0,0 +1,712 @@
...
+  // ... other props ...
+  goBack={goBack}
+/>
+```
+
+### Update the chat header
+
+In the chat header of our current team chat, we're already using a back button on smaller screens to go back to the conversation list:
+
+!! 2-back-button.jpg
```
Comment (keerlu) [inline how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md:471]: Reminder to myself to add the image on Ghost

File: how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md [hunk 3]
```diff
@@ -0,0 +1,712 @@
...
+    </div>
+  </ActionLink>
+</div>
+```
+
+Finally, we'll add some styling. See [our GitHub repo](https://github.com/talkjs/talkjs-examples/tree/master/react/team-chat-with-threads/theme/UserMessage.txt) for the full **UserMessage** template with added styles.
+
+You should now have something that looks like this:
+
+!! 3-team-chat.jpg
```
Comment (keerlu) [inline how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md:690]: Reminder to myself to add the image on Ghost

**Review by csmeyns — APPROVED** [review]
Awesome work! Clear steps and the structure works fine. Caveat: I myself didn't quite manage to get the reply counter to work 😕 but I attribute that to my being not too familiar with working with webhooks, so I must've done something wrong in setting things up there. Added a couple of small suggestion comments with an eye on the writing

File: how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md [hunk 4]
```diff
@@ -0,0 +1,712 @@
...
+
+In this section, we'll create a **Reply in thread** [message action](/Features/Message_Features/Message_Actions/) that creates a new thread for that message, or takes you to the thread if it already exists.
+
+The thread will be a standard TalkJS [conversation](/Reference/Concepts/Conversations/), with the message you're replying to duplicated at the top as a [system message](/Reference/Concepts/System_Messages/).
+
+### Replace the "Reply" action with a custom "Reply in thread" action
+
+First we'll remove the standard **Reply** message action and create a custom **Reply in thread** action instead. To do this:
+
+1.  Go to the **Roles** tab of the TalkJS dashboard.
```
[the remark this answers is not in the document]
Comment (keerlu) [reply how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md:90]: Oh yeah good point, thanks!

File: how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md [hunk 5]
```diff
@@ -0,0 +1,712 @@
...
+## Add a back button
+
+Currently, there's no way of getting back to the parent conversation from a thread. In this section, we'll fix this by updating the behavior of the left arrow icon button in the chat header.
+
+Some steps are similar to the steps for setting up the **Reply** message action in the last section, so we'll move more quickly through these this time.
+
+### Track the conversation history
+
+Our **Reply in thread** feature allows nested threads: you can reply to a message in one of the original conversations in the team chat (a channel or DM conversation), and then reply to a message in that thread, and so on. We'll use a stack to keep track of the history of previous conversations you've viewed at your current depth in the nesting.
+
```
[the remark this answers is not in the document]
Comment (keerlu) [reply how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md:365]: We looked at only allowing one level deep, but it's awkward in TalkJS. We don't have that much control over the message actions, so we can't have "Reply in thread" appear in the menu on the top level only. You could maybe do it all outside the message actions, but that would have been more complicated to do, so we stuck with the nested replies. Some discussion here: https://talkjs.slack.com/archives/C03507KTXJ4/p1725629686006659

I actually like nested threads, and would enjoy this option on other apps! But agree they are non-standard and we'd avoid them too if it was easier to do.

File: how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md [hunk 6]
```diff
@@ -0,0 +1,712 @@
...
+          d="M41.4157 12.5837C41.6019 12.7695 41.7497 12.9902 41.8505 13.2331C41.9513 13.4761 42.0032 13.7366 42.0032 13.9997C42.0032 14.2627 41.9513 14.5232 41.8505 14.7662C41.7497 15.0092 41.6019 15.2299 41.4157 15.4157L24.8277 31.9997L41.4157 48.5837C41.7912 48.9592 42.0022 49.4686 42.0022 49.9997C42.0022 50.5308 41.7912 51.0401 41.4157 51.4157C41.0401 51.7912 40.5308 52.0022 39.9997 52.0022C39.4686 52.0022 38.9592 51.7912 38.5837 51.4157L20.5837 33.4157C20.3974 33.2299 20.2496 33.0092 20.1488 32.7662C20.048 32.5232 19.9961 32.2627 19.9961 31.9997C19.9961 31.7366 20.048 31.4761 20.1488 31.2331C20.2496 30.9902 20.3974 30.7695 20.5837 30.5837L38.5837 12.5837C38.7695 12.3974 38.9902 12.2496 39.2331 12.1488C39.4761 12.048 39.7366 11.9961 39.9997 11.9961C40.2627 11.9961 40.5232 12.048 40.7662 12.1488C41.0092 12.2496 41.2299 12.3974 41.4157 12.5837Z"
+          fill="currentColor"
+        ></path>
+      </svg>
+    </div>
+  </ActionLink>
+</div>
+```
+
+Finally, we'll add some styling. See [our GitHub repo](https://github.com/talkjs/talkjs-examples/tree/master/react/team-chat-with-threads/theme/UserMessage.txt) for the full **UserMessage** template with added styles.
```
[the remark this answers is not in the document]
Comment (keerlu) [reply how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md:686]: I'd normally add the snippet, but in the case the diff from the normal team chat styling is pretty spaced out over several parts of the the UserMessage styles and would get pretty long and fiddly to include, so I just linked to it

File: how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md [hunk 7]
```diff
@@ -0,0 +1,712 @@
...
+
+```sh
+npm install express cors
+```
+
+As well as Express, this installs the `cors` package, which we use to set up CORS support on the server.
+
+```js
+import express from "express";
+import cors from "cors";
```
[the remark this answers is not in the document]
Comment (keerlu) [reply how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md]: Done!

File: how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md [hunk 8]
```diff
@@ -0,0 +1,712 @@
...
+app.listen(3001, () => console.log("Server is up"));
+```
+
+Replace `<APP_ID>` and `<SECRET_KEY>` with the values in the **Settings** page of your [TalkJS dashboard](https://talkjs.com/dashboard).
+
+Now, run the server with:
+
+```sh
+node server.js
+```
```
[the remark this answers is not in the document]
Comment (keerlu) [reply how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md]: Ah yes, thanks for catching that. I was wondering if thinking I could get away without without the `package.json` in the tutorial but forgot about `"type": "module"`. Added this in

File: how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md [hunk 9]
```diff
@@ -0,0 +1,712 @@
...
+    event.message.body,
+    Object.keys(event.message.conversation.participants)
+  );
+
+  // ...
+};
+```
+
+This posts the message ID, conversation ID, message text and a list of participants to a `/newThread` endpoint in the backend server you created earlier. Let's create the endpoint next:
+
```
[the remark this answers is not in the document]
Comment (keerlu) [reply how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md]: Done

File: how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md [hunk 10]
```diff
@@ -0,0 +1,712 @@
...
+First we'll remove the standard **Reply** message action and create a custom **Reply in thread** action instead. To do this:
+
+1.  Go to the **Roles** tab of the TalkJS dashboard.
+2.  Select the "default" role.
+3.  In **Actions and permissions** > **Built-in message actions**, set **Reply** to **None**.
+4.  In **Actions and permissions** > **Custom message actions**, create a new action with a name of "replyInThread" and a label of "Reply in thread", available to all messages for users with write permission.
+
+### Add an event handler for the "Reply in thread" action
+
+To listen for the new `replyInThread` event in your frontend code, pass the [`onCustomMessageAction` prop](https://talkjs.com/docs/Reference/React_SDK/Components/Chatbox/#event-props) to your `Chatbox`component:
```
[the remark this answers is not in the document]
Comment (keerlu) [reply how-to-add-threads-to-your-team-chat/how-to-add-threads-to-your-team-chat-with-talkjs.md]: Hm, I think I was indeed wanting to leave this open for that reason, but then I've ended up referring to `TeamChat.jsx` later on anyway. So I've put that in as you suggest, thanks

**Pull request merged**